lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,241 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstractions for system's DNF package information and repositories."""
+
+import os
+import re
+import shutil
+import subprocess
+from enum import Enum
+from typing import Dict, Optional, Union
+
+
+class Error(Exception):
+    """Raise when dnf encounters an execution error."""
+
+
+class _PackageState(Enum):
+    INSTALLED = "installed"
+    AVAILABLE = "available"
+    ABSENT = "absent"
+
+
+class PackageInfo:
+    """Dataclass representing DNF package information."""
+
+    def __init__(self, data: Dict[str, Union[str, _PackageState]]) -> None:
+        self._data = data
+
+    @property
+    def installed(self) -> bool:
+        """Determine if package is marked 'installed'."""
+        return self._data["state"] == _PackageState.INSTALLED
+
+    @property
+    def available(self) -> bool:
+        """Determine if package is marked 'available'."""
+        return self._data["state"] == _PackageState.AVAILABLE
+
+    @property
+    def absent(self) -> bool:
+        """Determine if package is marked 'absent'."""
+        return self._data["state"] == _PackageState.ABSENT
+
+    @property
+    def name(self) -> str:
+        """Get name of package."""
+        return self._data["name"]
+
+    @property
+    def arch(self) -> Optional[str]:
+        """Get architecture of package."""
+        return self._data.get("arch", None)
+
+    @property
+    def epoch(self) -> Optional[str]:
+        """Get epoch of package."""
+        return self._data.get("epoch", None)
+
+    @property
+    def version(self) -> Optional[str]:
+        """Get version of package."""
+        return self._data.get("version", None)
+
+    @property
+    def full_version(self) -> Optional[str]:
+        """Get full version of package."""
+        if self.absent:
+            return None
+
+        full_version = [self.version, f"-{self.release}"]
+        if self.epoch:
+            full_version.insert(0, f"{self.epoch}:")
+
+        return "".join(full_version)
+
+    @property
+    def release(self) -> Optional[str]:
+        """Get release of package."""
+        return self._data.get("release", None)
+
+    @property
+    def repo(self) -> Optional[str]:
+        """Get repository package is from."""
+        return self._data.get("repo", None)
+
+
+def version() -> str:
+    """Get version of `dnf` executable."""
+    return _dnf("--version").splitlines()[0]
+
+
+def installed() -> bool:
+    """Determine if the `dnf` executable is available on PATH."""
+    return shutil.which("dnf") is not None
+
+
+def update() -> None:
+    """Update all packages on the system."""
+    _dnf("update")
+
+
+def upgrade(*packages: str) -> None:
+    """Upgrade one or more packages.
+
+    Args:
+        *packages (str): Packages to upgrade on system.
+    """
+    if not packages:
+        _dnf("upgrade")
+    else:
+        _dnf("upgrade", *packages)
```

Comment:
I believe you don't need the `if` at all now and can just do `_dnf("upgrade", *packages)`, because Python doesn't mind passing an empty list as a `*args` list.

Reply:
That said, looking at what we have in the `install` and `remove` methods, we raise an explicit error if no packages were specified. So I'd recommend we have two methods here to avoid someone accidentally upgrading all packages when they meant "upgrade none":

```python
def upgrade_all():
    _dnf("upgrade")

def upgrade(*packages):
    if not packages:
        raise TypeError("No packages specified.")
    _dnf("upgrade", *packages)
```

Reply:
Hmm... I need to think about this. Would folks be potentially confused if `dnf.upgrade()` did not upgrade the system and rather threw an error at them? Reason saying is that folks who know `dnf` are going to expect that `dnf.upgrade()` will upgrade the whole system if they pass no arguments. Provided [Jon's reasoning](https://github.com/canonical/operator-libs-linux/pull/75#discussion_r1127102180) about making the dnf charm library feel natural to those familiar with it, would `upgrade_all(...)` be a break from what is expected? Let me know your thoughts. Maybe we just include a note in the docstring the clearly calls out that omitting packages will upgrade the whole system.

Reply:
Yeah, that's fair. I don't have strong feelings either way -- making it follow `dnf upgrade` closely and documenting it seems very reasonable.

Reply:
I decided to keep it as just upgrade in https://github.com/canonical/operator-libs-linux/pull/75/commits/d19923ec4e869150f7e257a6c55d8fcffffdaaf5. If people complain, we can add in a `upgrade_all()` method in a later pull request :sweat_smile: